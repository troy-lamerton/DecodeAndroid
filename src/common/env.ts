import { config } from 'dotenv';
import * as path from 'path';

config({ path: path.resolve(process.cwd(), '.env.local') });
config();

export interface IGlobals {
    TESTING: boolean;
    SYMBOLS_S3_BUCKET: string;
    SYMBOLS_S3_ACCESS_KEY: string;
    SYMBOLS_S3_SECRET_ACCESS_KEY: string;
}

export const Globals: IGlobals = {
    TESTING: Boolean(global['TESTING']), // only set by jest config, not env args
    SYMBOLS_S3_BUCKET: String(global['SYMBOLS_S3_BUCKET'] || process.env['SYMBOLS_S3_BUCKET']),
    SYMBOLS_S3_ACCESS_KEY: String(global['SYMBOLS_S3_ACCESS_KEY'] || process.env['SYMBOLS_S3_ACCESS_KEY']),
    SYMBOLS_S3_SECRET_ACCESS_KEY: String(global['SYMBOLS_S3_SECRET_ACCESS_KEY'] || process.env['SYMBOLS_S3_SECRET_ACCESS_KEY']),
};
