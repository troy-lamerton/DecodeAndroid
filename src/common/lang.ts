import { isUndefined, get, negate, isEmpty } from 'lodash';

export function callElseThrow<T>(func: undefined | (() => T), throws: any) {
    if (isUndefined(func)) {
        throw throws;
    }
    return func();
}

export const TODO = (msg?: string): never => {
    throw `[TODO] ${msg || 'not yet implemented'}`;
};

export function randomHex(n: number): string {
    let rs = '';
    let r = n % 8,
        q = (n - r) / 8,
        i;
    for (i = 0; i < q; i++) {
        rs += Math.random()
            .toString(16)
            .slice(2);
    }
    if (r > 0) {
        rs += Math.random()
            .toString(16)
            .slice(2, i);
    }
    return rs;
}

export const notEmpty = negate(isEmpty);

export const isDefined = <T>(value: any): value is T => value !== undefined;
