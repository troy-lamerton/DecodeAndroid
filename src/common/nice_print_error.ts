export function prettyPrintError(err: any): string {
    const output = serializeError(err, 1000);
    console.error('[ERROR]', output);
    return output;
}

export function serializeError(value: any, maxLength = 2048): string {
    if (typeof value === 'string') {
        return value.slice(0, maxLength);
    }

    if (typeof value === 'object') {
        return JSON.stringify(destroyCircular(value, [])).slice(0, maxLength);
    }

    // People sometimes throw things besides Error objectsâ€¦
    if (typeof value === 'function') {
        // JSON.stringify discards functions. We do too, unless a function is thrown directly.
        return `[Function: ${value.name || 'anonymous'}]`.slice(0, maxLength);
    }

    return value;
}

// @See(https://github.com/sindresorhus/serialize-error/blob/master/index.js)
const destroyCircular = (from: any, seen: any): object => {
    const to = Array.isArray(from) ? [] : {};

    seen.push(from);

    // TODO: Use `Object.entries() when targeting Node.js 8
    for (const key of Object.keys(from)) {
        const value = from[key];

        if (typeof value === 'function') {
            continue;
        }

        if (!value || typeof value !== 'object') {
            to[key] = value;
            continue;
        }

        if (!seen.includes(from[key])) {
            to[key] = destroyCircular(from[key], seen.slice());
            continue;
        }

        to[key] = '[Circular]';
    }

    const commonProperties = ['name', 'message', 'stack', 'code'];

    for (const property of commonProperties) {
        if (typeof from[property] === 'string') {
            to[property] = from[property];
        }
    }

    return to;
};
