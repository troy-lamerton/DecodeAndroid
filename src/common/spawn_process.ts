import { onExit, readableToString, streamEnd, streamWrite } from '@rauschma/stringio';
import { spawn, spawnSync } from 'child_process';
import { Writable } from 'stream';
import l from './logger';

/**
 * test me
 * Line A: We ignore stdin, want to access stdout via a stream and forward stderr to process.stderr.
 * Line B: We await until echoReadable() is completely done. Without this await, DONE would be printed before the first line of source.stdout.
 */
export async function demo_readProcessOutput(inputFile: string): Promise<string> {
    const source = spawn('cat', [inputFile], { stdio: ['ignore', 'pipe', process.stderr] }); // (A)

    return await readableToString(source.stdout!);
}

export async function demo_writeProcessInput(linesToWrite = ['First line', 'Second line']) {
    const sink = spawn('cat', [], { stdio: ['pipe', process.stdout, process.stderr] }); // (A)

    await writeToWritable(sink.stdin!, linesToWrite); // (B)
    await onExit(sink);
}

export function runProcessForOutput(executableFile: string, args: string[]): [string, string] {
    l.info(`running: ${executableFile} ${args.join(' ')}`);
    const proc = spawnSync(executableFile, args, { stdio: ['ignore', 'pipe', 'pipe'] });

    const code = proc.status;

    const errorOutput = proc.stderr.toString().trim();

    if (code === 0) {
        if (errorOutput) l.w(errorOutput);
        return [proc.stdout.toString().trimRight(), errorOutput];
    } else {
        throw errorOutput;
    }
}

export async function writeToProcessForOutput(
    executableFile: string,
    args: string[],
    inputLines: string[],
): Promise<string> {
    l.trace(`${executableFile} ${args.join(' ')}`);

    const proc = spawn(executableFile, args, { stdio: ['pipe', 'pipe', process.stderr] });

    l.trace('writing ' + inputLines.join('\n'));

    await writeToWritable(proc.stdin!, inputLines, true);
    l.trace('writing to child process done');
    const output = await readableToString(proc.stdout!);
    l.trace('read to end done');

    // if (!proc.killed) await onExit(proc);
    proc.kill();
    l.trace('process did exit done');

    return output.trimRight();
}

async function writeToWritable(writable: Writable, lines: string[], instantEnd = true) {
    for (const line of lines) {
        await streamWrite(writable, line + '\n');
    }
    if (instantEnd) await streamEnd(writable);
}
