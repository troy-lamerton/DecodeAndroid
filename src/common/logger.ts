import Pino, { Logger } from 'pino';
import { serializeError } from './nice_print_error';
import { Globals } from './env';

const tag = `crashdecoder_android`;

const level = process.env.LOG_LEVEL || 'debug';
// logging that conforms to cos-* elastic search index
const humanFormat: Pino.LoggerOptions = {
    base: {},
    prettyPrint: {
        translateTime: 'SYS:HH:MM:ss.l',
        colorize: true,
    },
    level,
};
const elasticFormat: Pino.LoggerOptions = {
    level,
    timestamp: () => `,"@timestamp":"${new Date().toJSON()}"`,
    changeLevelName: 'log_level',
    base: {
        tag,
    },
};

const printForHuman =
    process.env.HUMAN_LOGS || process.env.NODE_ENV === 'development' || Globals.TESTING;
const loggerConfig: Pino.LoggerOptions = printForHuman ? humanFormat : elasticFormat;

type Best = {
    w(msg: string, stat?: string): void;
    e(msg: any, stat?: string): void;
};

const l = Pino(loggerConfig) as (Logger & Best);

const defaults = (stat: string, logtype: string, merger: any = {}) => {
    return printForHuman
        ? {
              ...merger,
          }
        : {
              stat,
              logtype,
              ...merger,
          };
};

l.w = (msg: string, stat?: string) => {
    l.warn(defaults(stat || 'log', 'Warning', { msg }));
};

l.e = (err: any, stat?: string) => {
    l.error(
        defaults(stat || 'log', 'Error', {
            is_error: true,
            msg: serializeError(err),
        }),
    );
};

export default l;
