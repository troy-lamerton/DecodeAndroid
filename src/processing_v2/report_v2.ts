import {
    CrashReportV2Body,
    CommitSha,
    UnityPlatform,
    BuildEnvironmentE,
} from '../schema/crashes_v2/crashes_v2.schema';
import { DateTime, Duration } from 'luxon';
import { serializeError } from '../common/nice_print_error';
import { filter, isEmpty, isUndefined, size } from 'lodash';
import l from '../common/logger';
import { getSymbolsS3Key, IdleGameAndroidBuildArchE } from './download_symbols';
import { isDefined, notEmpty } from '../common/lang';
import { strict as assert } from 'assert';
import { AndroidNativeCrash } from './android_backtrace';

type ProcessingError = string;

export type ReportStatus =
    | 'init'
    | 'processing'
    | 'success' // everything went well
    | 'failed' // something went wrong during processing
    | 'faulted'; // a fatal error was thrown

export class CrashReportV2 {
    readonly startedProcessingAt: DateTime;
    /** Current status, set by methods in steps.ts */
    status: ReportStatus;
    /** Value is always available during processing.
     * To release memory, request_body is deleted once this report is safely in elastic search */
    request_body?: CrashReportV2Body;
    processing_job?: Promise<any>;
    readonly platform: UnityPlatform;
    readonly hash: string;
    private readonly errors: ProcessingError[] = [];
    readonly release_stage: BuildEnvironmentE;
    readonly client_commit: CommitSha;
    readonly client_version: string;
    readonly unity_build_id: string;
    readonly architecture: IdleGameAndroidBuildArchE | null;

    constructor(status: ReportStatus, request_body: CrashReportV2Body, hash: string) {
        this.startedProcessingAt = DateTime.utc();
        this.hash = hash;

        // copy common values from `request_body`
        // so they are available always, even when request_body is deleted
        const { platform, report_contents } = request_body;
        this.platform = platform;
        this.request_body = request_body;
        this.status = status;
        this.client_commit = request_body.client_stats.client_commit;
        this.client_version = request_body.client_stats.client_version;
        this.unity_build_id = request_body.unity_build_id;

        // computed values
        this.release_stage = request_body.index;
        try {
            this.architecture = AndroidNativeCrash.fromBugsnagJson(report_contents).architecture;
        } catch {
            this.architecture = null;
        }
    }

    getSymbolsDownloadKey(): string {
        if (!this.architecture) throw 'do not attempt to download symbols when architecture is unknown!';
        return getSymbolsS3Key({
            architecture: this.architecture,
            unityBuildId: this.unity_build_id,
        });
    }

    /**
     * Finalizes this report and releases allocations that were only needed for processing.
     * @param error
     * final error to add to this report.
     * @param faulted
     * this report is kaput, nothing more can be done with it.
     */
    markDone(error?: any, faulted?: boolean) {
        this.addError(error);

        if (faulted) {
            this.status = 'faulted';
            return;
        }

        if (isUndefined(error)) {
            this.status = 'success';
        } else {
            this.status = 'failed';
        }
        l.info(this.status + ' - finished processing report');
        if (this.errorCount)
            l.warn('Finished processing report with errors:\n' + this.errorOutputSecret);
    }

    deleteRequestBody() {
        delete this.request_body;
    }

    addError(throwable: any, logItNow?: boolean) {
        const error = serializeError(throwable, 512);
        l.debug('addError ' + error);

        // should not be adding new errors to a finalized report
        assert.notEqual(this.isFinalized(), true, 'this report is not finalized');

        if (isDefined(throwable)) {
            this.errors.push(error);
            if (logItNow) l.e(error);
        }
    }

    private isFinalized(): boolean {
        const finalStates: ReportStatus[] = ['failed', 'faulted', 'success'];
        return this.status in finalStates;
    }

    // info string is cached because request_body will be deleted at some point
    private _bodyInfos?: string;

    get bodyInfos(): string {
        if (this._bodyInfos) return this._bodyInfos;
        if (isEmpty(this.request_body)) return '<request_body is empty>';

        const body = this.request_body!;
        const { client_stats } = body;

        const infos = [
            `Build: ${body.index} - ${body.platform}`,
            `Version: ${client_stats.client_version} (commit: ${client_stats.client_commit})`,
            `Person: ${client_stats.person_id}`,
        ];

        return (this._bodyInfos = infos.join('\n'));
    }

    get errorCount(): number {
        return size(this.errors);
    }

    /**
     * May contain confidential information
     */
    get errorOutputSecret(): string | undefined {
        return isEmpty(this.errors)
            ? undefined
            : this.errors.map((str, i) => `#ERROR_${i + 1} ${str}`).join('\n');
    }

    /**
     * May contain confidential information
     */
    get descriptionSecret(): string {
        return filter([this.status, this.errorOutputSecret, this.bodyInfos], notEmpty).join(
            '\n',
        );
    }

    get description(): string {
        return filter(
            [this.status, `${this.errorCount} processing errors`, this.bodyInfos],
            notEmpty,
        ).join('\n');
    }
}
