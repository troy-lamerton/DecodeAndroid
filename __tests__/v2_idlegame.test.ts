import '../src/common/env';
import path from 'path';
import fs from 'fs-extra';
import {
    convertReportToDecodableText,
    decodeBacktrace,
    convertJvmReportToReadableStacktrace,
} from '../src/processing_v2/steps';
import { CrashReportV2 } from '../src/processing_v2/report_v2';
import { CrashReportV2Body } from '../src/schema/crashes_v2/crashes_v2.schema';
import {
    getUnityEngineSymbols,
    IdleGameAndroidBuildArchE,
    libil2cppDotSo,
    prepareUnityEngineSymbols,
    unity2019_2_8f1
} from '../src/processing_v2/download_symbols';

describe('idlegame crashes', () => {

    const dummy = {
        // native crash from an arm64 build
        idlegame_report_arm64_native: {
            files: {
                request_body_json: path.join(__dirname, 'dummy_data', 'requestBody_2_native.json'),
                // when you have access, these symbols can be downloaded from:
                // https://s3.console.aws.amazon.com/s3/buckets/gitlab-test-symbols/symbols/android/f39fefb7-01d9-4509-8f76-4a99957dcf24/ARM64/?region=us-east-2
                symbols_libil2cpp: path.join(__dirname, 'dummy_data', 'test_symbols', 'libil2cpp.so.debug'),
            }
        },
        idlegame_report_arm64_jvm: {
            files: {
                request_body_json: path.join(__dirname, 'dummy_data', 'requestBody_1_jvm.json'),
                symbols_libil2cpp: path.join(__dirname, 'dummy_data', 'test_symbols', 'libil2cpp.so.debug'),
            }
        }
    };

    describe('process report', () => {

        const tempFolder = path.join(__dirname, 'tmp');

        async function getPayloadContents(): Promise<string> {
            const jsonFileContents = await fs.readFile(dummy.idlegame_report_arm64_native.files.request_body_json);
            return jsonFileContents.toString();
        }

        test('parse report json', async () => {
            const payloadBodyJson = await getPayloadContents();
            const payloadBody = JSON.parse(payloadBodyJson) as CrashReportV2Body;
            const report = new CrashReportV2('init', payloadBody, 'unique-hash-here123');

            expect(report.hash).toBe('unique-hash-here123')
            expect(report.unity_build_id).toBe('f39fefb7-01d9-4509-8f76-4a99957dcf24')
        });

        test('s3 download key', async () => {
            const payloadBodyJson = await getPayloadContents();
            const payloadBody = JSON.parse(payloadBodyJson) as CrashReportV2Body;
            const report = new CrashReportV2('init', payloadBody, 'unique-hash-here123');
            const unityBuildId = 'f39fefb7-01d9-4509-8f76-4a99957dcf24';
            expect(report.unity_build_id).toBe(unityBuildId)
            expect(report.architecture).toBe(IdleGameAndroidBuildArchE.ARM64)
            const s3Key = report.getSymbolsDownloadKey()
            expect(s3Key).toBe(`symbols/android/${unityBuildId}/ARM64/libil2cpp.so.debug.gz`)
        })

        test('prepare engine symbol files', async () => {
            const payloadBodyJson = await getPayloadContents();
            const payloadBody = JSON.parse(payloadBodyJson) as CrashReportV2Body;
            const report = new CrashReportV2('init', payloadBody, 'unique-hash-here123');

            // copy in the unity sym files
            await prepareUnityEngineSymbols(tempFolder, report);

            // expect the correct unity symbol files
            const unitySymbols = await getUnityEngineSymbols(unity2019_2_8f1, IdleGameAndroidBuildArchE.ARM64);
            expect(unitySymbols.length).toBe(2);
            unitySymbols.forEach(filepath => {
                expect(filepath).toMatch(/^.+\/lib(?:main|unity).so$/i);
            });
        });

        test('get stacktrace from payload', async () => {
            const payloadBodyJson = await getPayloadContents();
            const payloadBody = JSON.parse(payloadBodyJson) as CrashReportV2Body;
            const { report_contents } = payloadBody;
            const decodableText = convertReportToDecodableText(report_contents);
            expect(decodableText).toMatchSnapshot();
        });

        test('decode the stacktrace', async () => {
            const payloadBodyJson = await getPayloadContents();
            const payloadBody = JSON.parse(payloadBodyJson) as CrashReportV2Body;
            const { report_contents } = payloadBody;
            const decodableText = convertReportToDecodableText(report_contents);
            const report = new CrashReportV2('init', payloadBody, 'unique-hash-here123');

            // copy over the app sym file (normally downloaded from s3 or copied from cache folder)
            const demoSymbolsFile = dummy.idlegame_report_arm64_native.files.symbols_libil2cpp;
            await fs.copy(demoSymbolsFile, path.join(tempFolder, libil2cppDotSo));

            // copy in the unity sym files
            await prepareUnityEngineSymbols(tempFolder, report);

            const decoded = await decodeBacktrace(decodableText, tempFolder);

            expect(decoded).toBeTruthy();
            expect(decoded).toMatchSnapshot();
        }, 40 * 1000);

        beforeAll(() => {
            fs.ensureDir(tempFolder);
        });

        afterAll(async () => {
            // delete files created during tests
            await fs.remove(tempFolder);
        });
    });

    describe('process jvm report', () => {
        async function getPayloadContents(): Promise<string> {
            const jsonFileContents = await fs.readFile(dummy.idlegame_report_arm64_jvm.files.request_body_json);
            return jsonFileContents.toString();
        }

        test('produce readable output', async () => {
            const payloadBodyJson = await getPayloadContents();
            const payloadBody = JSON.parse(payloadBodyJson) as CrashReportV2Body;
            const { report_contents } = payloadBody;
            const readableStacktrace = convertJvmReportToReadableStacktrace(report_contents);
            expect(readableStacktrace).toMatchSnapshot();
        })
    })

});
