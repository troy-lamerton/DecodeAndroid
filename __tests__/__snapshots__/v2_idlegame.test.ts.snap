// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`idlegame crashes process jvm report produce readable output 1`] = `
"#00 [vdso]. in _unknown_file:0
#01 libc.tgkill in tgkill:4
#02 libunity. in _unknown_file:0
#03 libil2cpp. in _unknown_file:0
#04 libil2cpp. in _unknown_file:0
#05 libil2cpp. in _unknown_file:0
#06 libil2cpp. in _unknown_file:0
#07 libil2cpp. in _unknown_file:0
#08 libil2cpp. in _unknown_file:0
#09 libil2cpp. in _unknown_file:0
#10 libil2cpp. in _unknown_file:0
#11 libil2cpp. in _unknown_file:0
#12 libil2cpp. in _unknown_file:0
#13 libil2cpp. in _unknown_file:0
#14 libil2cpp. in _unknown_file:0
#15 libil2cpp. in _unknown_file:0
#16 libil2cpp. in _unknown_file:0
#17 libil2cpp. in _unknown_file:0
#18 libil2cpp. in _unknown_file:0
#19 libil2cpp. in _unknown_file:0
#20 libil2cpp. in _unknown_file:0
#21 libil2cpp. in _unknown_file:0
#22 libunity. in _unknown_file:0
#23 libunity. in _unknown_file:0
#24 libunity. in _unknown_file:0
#25 libunity. in _unknown_file:0
#26 libunity. in _unknown_file:0
#27 libunity. in _unknown_file:0
#28 libunity. in _unknown_file:0
#29 libunity. in _unknown_file:0
#30 libunity. in _unknown_file:0
#31 base. in _unknown_file:0"
`;

exports[`idlegame crashes process report decode the stacktrace 1`] = `
"pid: 12300, tid: 12300, name: Thread  >>> app.app <<<
#00  6f128 /system/lib64/libc.so
#01 3c8750 <app-lib>/arm64/libunity.so: Routine void UnityGUID::Transfer<StreamedBinaryWrite>(StreamedBinaryWrite&) at ??:?
#02 2897e54 <app-lib>/arm64/libil2cpp.so: Routine RuntimeInvoker_TrueVoid_t22962CB4...017_Int32_t585191389E07734F19F3156FF88FB3EF4800D102(void (*)(), MethodInfo const*, void*, void**) at il2cppOutput/Il2CppInvokerTable.cpp:141169
#03 2b474e0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/Runtime.cpp:597
#04 2b4c4e8 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::vm::Runtime::InvokeConvertArgs(MethodInfo const*, void*, Il2CppObject**, int, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/Runtime.cpp:?
#05 2b273b0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::icalls::mscorlib::System::Reflection::MonoMethod::InternalInvoke(Il2CppReflectionMethod*, Il2CppObject*, Il2CppArray*, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\icalls\\\\mscorlib\\\\System.Reflection/MonoMethod.cpp:242
#06 5848c0c <app-lib>/arm64/libil2cpp.so: Routine MonoMethod_InternalInvoke_mB9735487...789 at il2cppOutput/mscorlib11.cpp:29139
#07 2fc41c0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp_codegen_get_virtual_invoke_data(unsigned int, Il2CppObject const*) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\codegen/il2cpp-codegen-il2cpp.h:305
#08 5a6efe8 <app-lib>/arm64/libil2cpp.so: Routine UnityAction_Invoke_mC9FF5AA1...71A at il2cppOutput/UnityEngine.CoreModule.cpp:28937
#09 5a4ebcc <app-lib>/arm64/libil2cpp.so: Routine UnityAction_1_Invoke_m3685D48F...92B_gshared at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/ClassInlines.h:?
#10 37ae3ec <app-lib>/arm64/libil2cpp.so: Routine InputService_ExecuteHandlerSafe_mF9232018...150 at il2cppOutput/Main82.cpp:61234
#11 37ae200 <app-lib>/arm64/libil2cpp.so: Routine InputService_HandleEvent_m31C7E8D1...5F6 at il2cppOutput/Main82.cpp:60725
#12 38cf348 <app-lib>/arm64/libil2cpp.so: Routine U3CU3Ec__DisplayClass0_0_U3C_ctorU3Eb__1_m76DF88E2...D14 at il2cppOutput/Main68.cpp:52575
#13 4b74634 <app-lib>/arm64/libil2cpp.so: Routine Action_1_Invoke_mB86FC1B3...1D5_gshared at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/ClassInlines.h:?
#14 514e354 <app-lib>/arm64/libil2cpp.so: Routine InlineExecuteAllSystem_Execute_m2A89F780...4CA at il2cppOutput/Helper.cpp:29406
#15 39f64d4 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::utils::LeaveTargetStack::push(int) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\utils/LeaveTargetStack.h:22
#16 3699b80 <app-lib>/arm64/libil2cpp.so: Routine il2cpp_codegen_get_virtual_invoke_data(unsigned int, Il2CppObject const*) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\codegen/il2cpp-codegen-il2cpp.h:305
#17 3c617d8 <app-lib>/arm64/libil2cpp.so: Routine ClientController_ControllerUpdate_m86390D47...BED at il2cppOutput/Main26.cpp:56324
#18 36998d8 <app-lib>/arm64/libil2cpp.so: Routine ControllerBridge_t316E85A0...F02::get_controller_4() const at il2cppOutput/Main37.cpp:34416
#19 2897c38 <app-lib>/arm64/libil2cpp.so: Routine RuntimeInvoker_TrueVoid_t22962CB4...017(void (*)(), MethodInfo const*, void*, void**) at il2cppOutput/Il2CppInvokerTable.cpp:141106
#20 2b474e0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/Runtime.cpp:597
#21 45187c <app-lib>/arm64/libunity.so: Routine void std::__adjust_heap<Experimental::GlobalIllumination::LightDataGI*, long, Experimental::GlobalIllumination::LightDataGI, __gnu_cxx::__ops::_Iter_comp_iter<SortByHashPred<Experimental::GlobalIllumination::LightDataGI, SceneLights::Hasher> > >(Experimental::GlobalIllumination::LightDataGI*, long, long, Experimental::GlobalIllumination::LightDataGI, __gnu_cxx::__ops::_Iter_comp_iter<SortByHashPred<Experimental::GlobalIllumination::LightDataGI, SceneLights::Hasher> >) at ??:?
#22 441cd8 <app-lib>/arm64/libunity.so: Routine GfxDeviceWorker::RunCommand(ThreadedStreamBuffer&) at ??:?
#23 43c6d4 <app-lib>/arm64/libunity.so: Routine std::_Rb_tree<GfxRasterState, std::pair<GfxRasterState const, ClientDeviceRasterState>, std::_Select1st<std::pair<GfxRasterState const, ClientDeviceRasterState> >, MemoryCompLess<GfxRasterState>, std::allocator<std::pair<GfxRasterState const, ClientDeviceRasterState> > >::_M_get_insert_unique_pos(GfxRasterState const&) at ??:?
#24 56f010 <app-lib>/arm64/libunity.so: Routine CachingManager::GetDefaultCache() at ??:?
#25 4cfd4c <app-lib>/arm64/libunity.so: Routine void std::__introsort_loop<int*, long, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter> >(int*, int*, long, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter>) at ??:?
#26 4cfd80 <app-lib>/arm64/libunity.so: Routine void std::__final_insertion_sort<int*, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter> >(int*, int*, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter>) at ??:?
#27 4cff54 <app-lib>/arm64/libunity.so: Routine void std::__adjust_heap<int*, long, int, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter> >(int*, long, long, int, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter>) at ??:?
#28 3e0ad4 <app-lib>/arm64/libunity.so: Routine void audio::mixer::GroupConstant::Transfer<SafeBinaryRead>(SafeBinaryRead&) at ??:?
#29 3da8f4 <app-lib>/arm64/libunity.so: Routine AudioSource::UpdateLocalChannels(dynamic_array<SoundChannel, 0ul>&) at ??:?
#30  648f0 /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/oat/arm64/base.odex"
`;

exports[`idlegame crashes process report get stacktrace from payload 1`] = `
"  #00  pc 6f128  /system/lib64/libc.so
  #01  pc 3c8750  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #02  pc 2897e54  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #03  pc 2b474e0  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #04  pc 2b4c4e8  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #05  pc 2b273b0  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #06  pc 5848c0c  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #07  pc 2fc41c0  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #08  pc 5a6efe8  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #09  pc 5a4ebcc  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #10  pc 37ae3ec  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #11  pc 37ae200  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #12  pc 38cf348  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #13  pc 4b74634  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #14  pc 514e354  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #15  pc 39f64d4  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #16  pc 3699b80  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #17  pc 3c617d8  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #18  pc 36998d8  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #19  pc 2897c38  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #20  pc 2b474e0  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libil2cpp.so
  #21  pc 45187c  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #22  pc 441cd8  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #23  pc 43c6d4  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #24  pc 56f010  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #25  pc 4cfd4c  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #26  pc 4cfd80  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #27  pc 4cff54  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #28  pc 3e0ad4  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #29  pc 3da8f4  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/lib/arm64/libunity.so
  #30  pc 648f0  /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/oat/arm64/base.odex"
`;
